shared/src/commonMain/kotlin/io/customer/shared/database/QueryHelper.kt

```diff
@@ -0,0 +1,264 @@
+     */
+    private fun runInTransactionAsync(
+        block: TransactionWithReturn<Unit>.() -> Unit,
+    ) = runSuspended { runInTransaction(block = block) }
+
+    /**
+     * Looks for unsent duplicate tasks that can be merged and may not be required to be sent as
+     * multiple events e.g. Identify, Add Device, etc.
+     */
+    @WithinTransaction
+    private fun Task.getDuplicateOrNull(): TrackingTask? = kotlin.runCatching {
+        return@runCatching if (activity.canBeMerged()) {
+            trackingTaskQueries.selectByType(
+                type = activity.type,
+                siteId = workspace.siteId,
+            ).executeAsOneOrNull()?.takeUnless { task ->
+                task.queueTaskStatus in listOf(
+                    QueueTaskStatus.QUEUED,
+                    QueueTaskStatus.SENDING,
+                    QueueTaskStatus.SENT,
+                )
+            }
+        } else null
+    }.getOrNull()
+
+    /**
+     * Merges activities that are merge-able, e.g. attributes, etc.
+     */
+    @WithinTransaction
+    private fun TrackingTask.mergeActivities(activity: Activity): Activity? {
+        var mergedActivity: Activity? = null
+        kotlin.runCatching {
+            activity.merge(other = jsonAdapter.fromJSON(Activity::class, activityJson))
+        }.fold(
+            onSuccess = { value -> mergedActivity = value },
+            onFailure = { ex ->
+                logger.fatal("Failed to parse activity ${type}, model version ${activityModelVersion}. Reason: ${ex.message}")
+            },
+        )
+        return mergedActivity
+    }
+
+    /**
+     * Updates status of tasks with the given values.
+     *
+     * For private use only, should only be called within database transaction.
+     */
+    @WithinTransaction
+    private fun updateTasksStatusInternal(
+        status: QueueTaskStatus,
+        tasks: List<TrackingTask>,
+    ): QueueTaskResult {
+        val taskIds = tasks.map { task -> task.uuid }
+        val result = kotlin.runCatching {
+            trackingTaskQueries.updateTasksStatus(
+                updatedAt = dateTimeUtil.now,
+                status = status,
+                ids = taskIds,
+                siteId = workspace.siteId,
+            )
+        }
+        result.onFailure { ex ->
+            logger.error("Unable to update status $status for tasks ${taskIds.joinToString(separator = ",")}. Reason: ${ex.message}")
+        }
+        return result.isSuccess
+    }
+
+    /**
+     * Inserts activity task safely to the database.
+     *
+     * For private use only, should only be called within database transaction.
+     */
+    @WithinTransaction
+    private fun insertTaskInternal(task: Task) = kotlin.runCatching {
+        val currentTime = dateTimeUtil.now
+        val duplicateTask = task.getDuplicateOrNull()
+        val mergedActivity = duplicateTask?.mergeActivities(activity = task.activity)
+        val activity = mergedActivity ?: task.activity
```

Comment:
Support for merging was added when the DB design was slightly different initially, but since I put some time and code to support this, I decided to keep it here for initial PR. The only scenario I think it can be helpful right now is when Add Device event is added multiple times, it can help exclude duplicate events by merging them. What pros/cons can you think for this? Should we keep/drop it?

Reply:
There is a lot of code in this function and seems to be somewhat complex. 

I would find it helpful to see this function be as simple as possible and add complexity as we see the need for it. I am not convinced that we need a "merging" feature until I see a PR that uses this code and shows the value in merging. 

Therefore, I suggest that we remove this "merging" feature in this PR and add it later when it's needed. 

I understand that you spent the time and effort to add the code here and removing it feels like a waste. However, you never lose git commits. You can delete all of the merging code, make a git commit, then come back to that git commit at a later time to reference it when you need to add the feature.

Reply:
Agreed. I tried to make it simpler but couldn't make it better than this. I too don't have a lot of use cases for this except the Add Device event, so I think it is a good idea to remove all changes in a single commit so we can reuse it in future if needed, will remove this.